### PR TITLE
validate iyk tap ref 

### PR DESCRIPTION
**Pull request by amit-shah-cb**
validate iykRef and update treasure-box with score*physical_tap_multiplier if tap is valid. 

This is currently a get request which would be triggered in a `useEffect` of a page that is served when the `iyk` nfc is tapped and `redirect-url` is opened.  The url has `iykRef` appended to it  

The page should make a `GET` request to `/api/treasure-box/iyk` with with the following params:
 userAddress=> from wallet address
 gameId => constant?
 iykRef=> from window.location.href 

This can be changed to a `POST` request trivially with the same params in body

### CODE REVIEW

File: src/app/api/profile/level/route.ts [hunk 1]
```diff
@@ -1,8 +1,8 @@
 import { NextRequest, NextResponse } from 'next/server';
 import '@/utils/helper';
-import { Badge, BadgeTypeEnum } from '../../../../hooks/types';
-
+import { BadgeTypeEnum } from '../../../../hooks/types';
```
Comment (amit-shah-cb) [inline src/app/api/profile/level/route.ts:3]: unused import

**Review by taycaldwell — APPROVED** [review]

File: src/app/api/levels/route.ts [hunk 2]
```diff
@@ -10,9 +11,9 @@ const supabase = createClient<Database>(
 
 export async function GET(request: NextRequest) {
   const searchParams = request.nextUrl.searchParams;
-  const gameId = searchParams.get('gameId');
+  const gameId = toBigInt(searchParams.get('gameId') as string);
 
-  if (!gameId) {
+  if (gameId == null) {
```
Comment (siwan-cb) [inline src/app/api/levels/route.ts:16]:
```suggestion
  if (gameId === null) {
```

File: src/app/api/profile/level/route.ts [hunk 3]
```diff
@@ -25,7 +25,15 @@ type QueryData = {
 export async function GET(req: NextRequest) {
   const searchParams = req.nextUrl.searchParams;
   const userAddress = searchParams.get('userAddress') as string;
-  const gameId = BigInt(searchParams.get('gameId') as string);
+  const gameId = toBigInt(searchParams.get('gameId') as string);
+  if (!userAddress || gameId == null) {
```
Comment (siwan-cb) [inline src/app/api/profile/level/route.ts:29]:
```suggestion
  if (!userAddress || gameId === null) {
```

File: src/app/api/profile/score/route.ts [hunk 4]
```diff
@@ -15,11 +14,17 @@ curl -X POST 'http://localhost:3000/api/level/claim' -d ' {"type":"INSERT","tabl
 */
 
 export async function GET(req: NextRequest) {
-  const headersList = headers();
-  const referer = headersList.get('x-app-secert');
   const searchParams = req.nextUrl.searchParams;
   const userAddress = searchParams.get('userAddress') as string;
-  const gameId = BigInt(searchParams.get('gameId') as string);
+  const gameId = toBigInt(searchParams.get('gameId') as string);
+  if (!userAddress || gameId == null) {
```
Comment (siwan-cb) [inline src/app/api/profile/score/route.ts:20]:
```suggestion
  if (!userAddress || gameId === null) {
```

File: src/app/api/profile/state/route.ts [hunk 5]
```diff
@@ -25,7 +26,15 @@ type QueryData = {
 export async function GET(req: NextRequest) {
   const searchParams = req.nextUrl.searchParams;
   const userAddress = searchParams.get('userAddress') as string;
-  const gameId = BigInt(searchParams.get('gameId') as string);
+  const gameId = toBigInt(searchParams.get('gameId') as string);
+  if (!userAddress || gameId == null) {
```
Comment (siwan-cb) [inline src/app/api/profile/state/route.ts:30]:
```suggestion
  if (!userAddress || gameId === null) {
```

File: src/app/api/treasure-box/iyk/route.ts [hunk 6]
```diff
@@ -1,16 +1,70 @@
...
+  process.env.PHYSICAL_TAP_MULTIPLIER as string
+) as number;
 
 export async function GET(request: NextRequest) {
   const searchParams = request.nextUrl.searchParams;
   const iykRef = searchParams.get('iykRef');
+  const userAddress = searchParams.get('userAddress');
+  const gameId = toBigInt(searchParams.get('gameId') as string);
+
+  if (!userAddress || gameId == null || !iykRef) {
```
Comment (siwan-cb) [inline src/app/api/treasure-box/iyk/route.ts:23]:
```suggestion
  if (!userAddress || gameId === null || !iykRef) {
```

File: src/app/api/treasure-box/iyk/route.ts [hunk 7]
```diff
@@ -1,16 +1,70 @@
...
+      `Missing parameters: userAddress: ${userAddress}, gameId: ${gameId}, iykRef: ${iykRef}`,
+      {
+        status: 400,
+      }
+    );
+  }
   const result = await getTapRef(iykRef as string);
-  console.log(result);
-  return NextResponse.json(result);
+  if (result == null || !result.isValidRef) {
```
Comment (siwan-cb) [inline src/app/api/treasure-box/iyk/route.ts:32]:
```suggestion
  if (result === null || !result.isValidRef) {
```

File: src/app/api/treasure-box/rank/route.ts [hunk 8]
```diff
@@ -16,9 +16,9 @@ const supabase = createClient(
 export async function GET(request: NextRequest) {
   const searchParams = request.nextUrl.searchParams;
   const userAddress = searchParams.get('userAddress');
-  const gameId = searchParams.get('gameId');
+  const gameId = toBigInt(searchParams.get('gameId') as string);
 
-  if (!userAddress || !gameId) {
+  if (!userAddress || gameId == null) {
```
Comment (siwan-cb) [inline src/app/api/treasure-box/rank/route.ts:21]:
```suggestion
  if (!userAddress || gameId === null) {
```

File: src/app/api/treasure-box/top-ranks/route.ts [hunk 9]
```diff
@@ -11,9 +12,9 @@ const supabase = createClient<Database>(
 // return top 10 ranks
 export async function GET(request: NextRequest) {
   const searchParams = request.nextUrl.searchParams;
-  const gameId = searchParams.get('gameId');
+  const gameId = toBigInt(searchParams.get('gameId') as string);
 
-  if (!gameId) {
+  if (gameId == null) {
```
Comment (siwan-cb) [inline src/app/api/treasure-box/top-ranks/route.ts:17]:
```suggestion
  if (gameId === null) {
```

File: src/utils/iyk.ts [hunk 10]
```diff
@@ -1,15 +1,36 @@
...
+}
 
-    /*
+export async function getTapRef(iykRef: string): Promise<TapRef | null> {
+  const data = await axios.get<TapRef>(`https://api.iyk.app/refs/${iykRef}`, {
+    headers: {
+      'x-iyk-api-key': process.env.IYK_API_KEY as string,
+    },
+  });
+  if (data.status != 200) {
```
Comment (siwan-cb) [inline src/utils/iyk.ts:30]:
```suggestion
  if (data.status !== 200) {
```

**Pull request merged**